Sources/Regex/RegexToken2.swift

```diff
@@ -0,0 +1,286 @@
+  //
+  // \n, \s, \Q, \b, \A, \K, ...
+  case escaped(EscapedSingle)
+
+  // A Unicode scalar value written as a literal
+  //
+  // \u{...}, \0dd, \x{...}, ...
+  case scalar(Unicode.Scalar)
+
+  // A Unicode property, category, or script
+  //
+  // \p{...}, \p{^...}, \P
+  case property(Prop, inverted: Bool)
+
+  // A control character
+  //
+  // \cx, \C-x, \M-x, \M-\C-x, ...
+  case control(Character)
+  case keyboardMeta(Character, alsoControl: Bool) // Oniguruma
+
+  // Post-fix quantification
+  case quantifier(Quantifier)
+
+  // A named set (using POSIX syntax)
+  //
+  // [[:...:]], [[:^...:]]
+  case named(POSIXSet, inverted: Bool)
+
+  // [, [^
+  case startCustomCharacterClass(inverted: Bool)
+
+  // ]
+  case endCustomCharacterClass
+
+  // The start of a group
+  case startGroup(GroupStart)
+
+  // )
+  case endGroup
+
+  // .
+  case any
+
+  // ^
+  case startOfLine
+
+  // $
+  case endOfLine
+
+  // |
+  case alternation
+
+  // References
+  case backreference(Reference)
+  case subpattern(Reference)
+  case condition(Reference)
+
+  //
+  // Below only present in custom character classes
+  //
+
+  // `-`, sometimes the range binary operator
+  //
+  // Lexer knows to emit .char("-") outside of custom char class.
+  //
+  // TODO: can lexer check positioning easily?
+  //
+  case minus
+
+  // &&, ~~, --, ||
+  case setOperator(SetOperator)
+
+}
+
+import Util
+
+extension NewToken {
+  typealias POSIXSet = Unicode.POSIXCharacterSet
+
+  // Characters, character types, literals, etc., derived from an escape
+  // sequence.
+  enum EscapedSingle: Character {
+    // Literal single characters
+    case alarm         = "a"
+    case escape        = "e"
+    case formfeed      = "f"
+    case newline       = "n"
+    case cariageReturn = "r"
+    case tab           = "t"
+
+    // Character types
+    case singleDataUnit          = "C"
+    case decimalDigit            = "d"
+    case notDecimalDigit         = "D"
+    case horizontalWhitespace    = "h"
+    case notHorizontalWhitespace = "H"
+    case notNewline              = "N"
+    case newlineSequence         = "R"
+    case whitespace              = "s"
+    case notWhitespace           = "S"
+    case verticalTab             = "v"
+    case notVerticalTab          = "V"
+    case wordCharacter           = "w"
+    case notWordCharacter        = "W"
+    case graphemeCluster         = "X"
+
+    // Quoting
+    case startQuote = "Q"
+    case endQuote   = "E"
+
+    // Assertions
+    case wordBoundary    = "b"
+    case notWordBoundary = "B"
+
+    // Anchors
+    case startOfSubject                 = "A"
+    case endOfSubjectBeforeNewline      = "Z"
+    case endOfSubject                   = "z"
+    case firstMatchingPositionInSubject = "G"
+
+    // Other
+    case resetStartOfMatch = "K"
+
+    // Oniguruma
+    case trueAnychar = "O"
+    case textSegnemt = "y"
+    case notTextSegment = "Y"
+
+  }
+
+  enum Prop {
+    case gc(Unicode.GeneralCategory)
```

Comment:
Just realized this doesn't cover the meta-categories e.g `L` which matches all the letter categories, should we spin off a separate type to handle those?

Reply:
I'm not sure how best to model/unify this. The `FlattendedOnigurumaUnicodeProperty` mega enum just puts them all flat. It makes sense to me that lexical analysis can resolve this, though.